projects/httpProxy.c

```diff
@@ -0,0 +1,164 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <netdb.h>
+
+#define PORT 80
+#define BUFFER_SIZE 4096
+
+void* handle_connection(void* client_socket);
+int connect_to_target(char* host, int port);
+
+int main(int argc, char* argv[]) {
+    int server_fd, client_socket;
+    struct sockaddr_in address;
+    int addr_len = sizeof(address);
+    int opt = 1;
+
+    // Создание сокета сервера
+    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
+        perror("socket failed");
+        exit(EXIT_FAILURE);
+    }
+
+    // Настройка сокета для повторного использования порта
+    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
+        perror("setsockopt");
+        exit(EXIT_FAILURE);
+    }
+
+    // Адрес сервера
+    address.sin_family = AF_INET;
+    address.sin_addr.s_addr = INADDR_ANY;
+    address.sin_port = htons(PORT);
+
+    // Привязка сокета к порту 80
+    if (bind(server_fd, (struct sockaddr*)&address, sizeof(address)) < 0) {
+        perror("bind failed");
+        exit(EXIT_FAILURE);
+    }
+
+    // Начало прослушивания
+    if (listen(server_fd, 10) < 0) {
+        perror("listen");
+        exit(EXIT_FAILURE);
+    }
+
+    printf("HTTP Proxy Server Listening on port %d\n", PORT);
+
+    while (1) {
+        // Принятие соединения от клиента
+        if ((client_socket = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addr_len)) < 0) {
+            perror("accept");
+            continue;
+        }
+
+        // Создание потока для обработки соединения
+        pthread_t thread_id;
+        if (pthread_create(&thread_id, NULL, handle_connection, (void*)&client_socket) != 0) {
```

Comment:
what about joining threads ?

Reply:
Can I use detach threads? If not, then I’ll think about storing and joining threads. (Simply calling Join all the time can greatly disrupt the efficiency of the program, creating an array, and if it overflows, go through the elements with a call to join - with a large flow of connections, we can hang on Join. Ideally, just implement a thread pool)

Reply:
Use any solution. We have limitations on amount of threads (lab 1)  and it is not desired to have the same issue here.

Reply:
fixed